src/llvm-propagate-addrspaces.cpp

```diff
@@ -74,67 +77,136 @@ static bool isSpecialAS(unsigned AS) {
     return AddressSpace::FirstSpecial <= AS && AS <= AddressSpace::LastSpecial;
 }
 
+void PropagateJuliaAddrspaces::PoisonValues(std::vector<Value *> &Worklist) {
+    while (!Worklist.empty()) {
+        Value *CurrentV = Worklist.back();
+        Worklist.pop_back();
+        for (Value *User : CurrentV->users()) {
+            if (Visited.count(User))
+                continue;
+            Visited.insert(CurrentV);
+            Worklist.push_back(User);
+        }
+    }
+}
+
 Value *PropagateJuliaAddrspaces::LiftPointer(Value *V, Type *LocTy, Instruction *InsertPt) {
     SmallVector<Value *, 4> Stack;
-    Value *CurrentV = V;
+    std::vector<Value *> Worklist;
+    std::set<Value *> LocalVisited;
+    Worklist.push_back(V);
     // Follow pointer casts back, see if we're based on a pointer in
     // an untracked address space, in which case we're allowed to drop
     // intermediate addrspace casts.
-    while (true) {
-        Stack.push_back(CurrentV);
-        if (isa<BitCastInst>(CurrentV))
-            CurrentV = cast<BitCastInst>(CurrentV)->getOperand(0);
-        else if (isa<AddrSpaceCastInst>(CurrentV)) {
-            CurrentV = cast<AddrSpaceCastInst>(CurrentV)->getOperand(0);
-            if (!isSpecialAS(getValueAddrSpace(CurrentV)))
-                break;
+    while (!Worklist.empty()) {
+        Value *CurrentV = Worklist.back();
+        Worklist.pop_back();
+        if (LocalVisited.count(CurrentV)) {
+            continue;
         }
-        else if (isa<GetElementPtrInst>(CurrentV)) {
-            if (LiftingMap.count(CurrentV)) {
-                CurrentV = LiftingMap[CurrentV];
+        while (true) {
+            if (isa<BitCastInst>(CurrentV))
+                CurrentV = cast<BitCastInst>(CurrentV)->getOperand(0);
+            else if (isa<AddrSpaceCastInst>(CurrentV)) {
+                CurrentV = cast<AddrSpaceCastInst>(CurrentV)->getOperand(0);
+                if (!isSpecialAS(getValueAddrSpace(CurrentV)))
+                    break;
+            }
+            else if (isa<GetElementPtrInst>(CurrentV)) {
+                if (LiftingMap.count(CurrentV)) {
+                    CurrentV = LiftingMap[CurrentV];
+                    break;
+                } else if (Visited.count(CurrentV)) {
+                    return nullptr;
+                }
+                Stack.push_back(CurrentV);
+                LocalVisited.insert(CurrentV);
```

Comment:
short-circuit is nice enough, but GEP can't participate in cycles. This is needed on PhiNode, not here

Reply:
Two values branches of a PhiNode can (and do) reference the same GetElementPtrInst though.